File: RGBLed/Swatch.cpp
#include "Swatch.h"
#include <Arduino.h>

Swatch::Swatch(const char color[4]){
  _red = (int) color[0];
  _green = (int) color[1];
  _blue = (int) color[2];
}

Swatch::Swatch(int red, int green, int blue){
  _red = red;
  _green = green;
  _blue = blue;
}

Swatch::Swatch(int color){
  this->fromInt(color);
}

Swatch::Swatch(){
  
}

int Swatch::getRed(){
  return _red;
}

int Swatch::getPWMRed(){
  return (int)map(_red, 0, 256, 0, 1024);
}

void Swatch::setRed(int red){
  _red = red;
}

int Swatch::getBlue(){
  return _blue;
}

int Swatch::getPWMBlue(){
  return (int)map(_blue, 0, 256, 0, 1024);
}

void Swatch::setBlue(int blue){
  _blue = blue;
}

int Swatch::getGreen(){
  return _green;
}

int Swatch::getPWMGreen(){
  return (int)map(_green, 0, 256, 0, 1024);
}

void Swatch::setGreen(int green){
  _green = green;
}


bool Swatch::colorEquals(Swatch other){
  return other.getRed() == _red && other.getGreen() == _green && other.getBlue() == _blue;
}

void Swatch::fromInt(int color){
  _red = (color & 0x00FF0000) >> 16;
  _green = (color & 0x0000FF00) >> 8;
  _blue = color & 0x000000FF;
}

int Swatch::asInt(){
  return (0xFF << 24) | (_red << 16) | (_green << 8) | _blue;
}

void Swatch::black(){
  _red = 0;
  _green = 0;
  _blue = 0;
}

void Swatch::white(){
  _red = 255;
  _green = 255;
  _blue = 255;
}

File: src/main/java/com/example/hexbow/lamp/Swatch.java
package com.example.hexbow.lamp;

import androidx.annotation.Nullable;

import com.example.hexbow.callback.Instructable;

public class Swatch implements Instructable {
    private int red, green, blue, alpha;

    public Swatch(){
        red = 0;
        green = 0;
        blue = 0;
        alpha = 0;
    }

    public Swatch(int red, int  green, int blue, int alpha){
        this.red = red;
        this.green = green;
        this.blue = blue;
        this.alpha = alpha;
    }

    public Swatch(Swatch other){
        this.red = other.red;
        this.green = other.green;
        this.blue = other.blue;
        this.alpha = other.alpha;
    }

    public int getRed() {
        return red;
    }

    public Swatch setRed(int red) {
        this.red = red;
        return this;
    }

    public int getGreen() {
        return green;
    }

    public Swatch setGreen(int green) {
        this.green = green;
        return this;
    }

    public int getBlue() {
        return blue;
    }

    public Swatch setBlue(int blue) {
        this.blue = blue;
        return this;
    }

    public int getAlpha() {
        return alpha;
    }

    public Swatch setAlpha(int alpha) {
        this.alpha = alpha;
        return this;
    }

    public int toInt() {

        //System.out.println(String.format("Alpha 0x%08X, Red 0x%08X, Green 0x%08X, Blue 0x%08X, Sum 0x%08x\n", alpha<<24, red<<16, green<<8, blue, (alpha << 24) | (red << 16) | (green << 8) | blue));

        return (alpha << 24) | (red << 16) | (green << 8) | blue;
    }

    public void fromInt(int color){
        alpha = (color & 0xFF000000) >> 24;
        red = (color & 0x00FF0000) >> 16;
        green = (color & 0x0000FF00) >> 8;
        blue = color & 0x000000FF;
    }

    public Swatch invert(){
        return new Swatch(255 - red, 255 - green, 255 - blue, alpha);
    }

    public Swatch grayScale(){

        int gray = (red + blue + green)/3;

        return new Swatch(gray, gray, gray, alpha);
    }

    public Swatch swatch(double factor){
        int nred = factor * red > 255 ? 255 : (int) (factor * red);
        int ngreen = factor * green > 255 ? 255 : (int) (factor * green);
        int nblue = factor * blue > 255 ? 255 : (int) (factor * blue);

        return new Swatch(nred, ngreen, nblue, alpha);

    }

    public Swatch swatch(int dColor){
        int nred = dColor + red;
        int ngreen = dColor + green;
        int nblue = dColor + blue;

        if(nred < 0) nred = 0;
        if(nred > 255) nred = 255;

        if(ngreen < 0) ngreen = 0;
        if(ngreen > 255) ngreen = 255;

        if(nblue < 0) nblue = 0;
        if(nblue > 255) nblue = 255;

        return new Swatch(nred, ngreen, nblue, alpha);
    }

    public String instruction(){
        return this.toInt() + "";
    }

    @Override
    public String toString(){
        return red + "," + green + "," + blue;
    }

    @Override
    public boolean equals(@Nullable Object obj) {
        Swatch other = (Swatch) obj;

        return other.red == red && other.green == green && other.blue == blue && other.alpha == alpha;
    }

    public static Swatch random(){
        return new Swatch((int)(Math.random() * 255),(int)(Math.random() * 255),(int)(Math.random() * 255),255);
    }

    public static Swatch black(){
        return new Swatch(0,0,0,255);
    }

    public static Swatch white(){
        return new Swatch(255,255,255,255);
    }
}
